spec/ddtrace/configuration/settings_spec.rb

```diff
@@ -203,164 +537,367 @@
     end
 
-    context 'given :debug' do
-      subject(:configure) { settings.tracer(debug: debug) }
+    describe '#sampler' do
+      subject(:sampler) { settings.tracer.sampler }
+      it { is_expected.to be nil }
+    end
 
-      shared_examples_for 'debug toggle' do
-        before { Datadog::Logger.debug_logging = !debug }
-        after { Datadog::Logger.debug_logging = false }
+    describe '#sampler=' do
+      let(:sampler) { instance_double(Datadog::PrioritySampler) }
 
-        it do
-          expect { configure }.to change { Datadog::Logger.debug_logging }
-            .from(!debug)
-            .to(debug)
-        end
+      it 'updates the #sampler setting' do
+        expect { settings.tracer.sampler = sampler }
+          .to change { settings.tracer.sampler }
+          .from(nil)
+          .to(sampler)
       end
+    end
+
+    describe '#transport_options' do
+      subject(:transport_options) { settings.tracer.transport_options }
+      it { is_expected.to eq({}) }
 
-      context 'as true' do
-        it_behaves_like 'debug toggle' do
-          let(:debug) { true }
+      context 'when modified' do
+        it 'does not modify the default by reference' do
+          settings.tracer.transport_options[:foo] = :bar
+          expect(settings.tracer.transport_options).to_not be_empty
+          expect(settings.tracer.options[:transport_options].default_value).to be_empty
         end
       end
+    end
 
-      context 'as false' do
-        it_behaves_like 'debug toggle' do
-          let(:debug) { false }
-        end
+    describe '#transport_options=' do
+      let(:options) { { hostname: 'my-agent' } }
+
+      it 'updates the #transport_options setting' do
+        expect { settings.tracer.transport_options = options }
+          .to change { settings.tracer.transport_options }
+          .from({})
+          .to(options)
       end
     end
 
-    context 'given some settings' do
-      let(:tracer) { Datadog::Tracer.new }
+    describe '#writer' do
+      subject(:writer) { settings.tracer.writer }
+      it { is_expected.to be nil }
+    end
+
+    describe '#writer=' do
+      let(:writer) { instance_double(Datadog::Writer) }
 
-      before do
-        settings.tracer(
-          enabled: false,
-          hostname: 'tracer.host.com',
-          port: 1234,
-          env: :config_test,
-          tags: { foo: :bar },
-          writer_options: { buffer_size: 1234 },
-          instance: tracer
-        )
+      it 'updates the #writer setting' do
+        expect { settings.tracer.writer = writer }
+          .to change { settings.tracer.writer }
+          .from(nil)
+          .to(writer)
       end
+    end
+
+    describe '#writer_options' do
+      subject(:writer_options) { settings.tracer.writer_options }
+      it { is_expected.to eq({}) }
 
-      after do
-        Datadog::Logger.debug_logging = false
+      context 'when modified' do
+        it 'does not modify the default by reference' do
```

Comment:
This was an issue when `option :writer_options, default: {}` was defined, because a user could errantly modify the default by reference.